c7n/resources/rds.py

```diff
@@ -274,6 +274,133 @@ def process(self, dbs):
+
+    def _list_versions(self, client, resource):
+        from distutils.version import StrictVersion as sv
+        versions = client.describe_db_engine_versions(
+            Engine=resource['Engine'],
+            EngineVersion=resource['EngineVersion'])['DBEngineVersions'][0]
+
+        version = resource['EngineVersion']
+        for v in versions['ValidUpgradeTarget']:
+            if sv(v['EngineVersion']) > sv(version):
+                version = v['EngineVersion']
+        return version
+
+    def process(self, resources, event=None):
+        client = local_session(self.manager.session_factory).client('rds')
+        results = []
+        for r in resources:
+            version = self._list_versions(client, r)
+            if version != r['EngineVersion']:
+                results.append(r)
+        return results
+
+
+@filters.register('upgrade-complete')
+class UpgradeComplete(Filter):
+
+    schema = type_schema('upgrade-complete')
+
+    def _list_versions(self, client, resource):
+        from distutils.version import StrictVersion as sv
+        versions = client.describe_db_engine_versions(
+            Engine=resource['Engine'],
+            EngineVersion=resource['EngineVersion'])['DBEngineVersions'][0]
+
+        version = resource['EngineVersion']
+        for v in versions['ValidUpgradeTarget']:
+            if sv(v['EngineVersion']) > sv(version):
+                version = v['EngineVersion']
+        return version
+
+    def process(self, resources, event=None):
+        client = local_session(self.manager.session_factory).client('rds')
+        results = []
+        for r in resources:
+            version = self._list_versions(client, r)
+            if version == r['EngineVersion']:
+                results.append(r)
+        return results
+
+
+@actions.register('set-minor-upgrade-flag')
+class SetMinorUpgradeFlag(BaseAction):
+
+    schema = type_schema('set-minor-upgrade-flag', value={'type': 'boolean'})
+
+    def _set_flag(self, client, resource):
+        client.modify_db_instance(
+            DBInstanceIdentifier=resource['DBInstanceIdentifier'],
+            AutoMinorVersionUpgrade=self.data.get('value', True))
+
+    def process(self, resources):
+        futures = []
+        client = local_session(self.manager.session_factory).client('rds')
+        with self.executor_factory(max_workers=3) as w:
+            for r in resources:
+                futures.append(w.submit(self._set_flag, client, r))
+        for f in as_completed(futures):
+            if f.exception():
+                self.log.error(
+                    "Exception setting upgrade flag \n %s",f.exception())
+
+
+@actions.register('upgrade-minor')
+class UpgradeMinorRDS(BaseAction):
+
+    schema = type_schema('upgrade-minor')
+
+    def _upgrade(self, client, resource):
+        from distutils.version import StrictVersion as sv
+
+        version = resource['EngineVersion']
+        versions = client.describe_db_engine_versions(
```

Comment:
this seems better as a filter annotation on the resource to get the upgrade if present.

also the repetitive implementation of this logic suggest it would be better as a function.